// Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package nvidia_resiliency_ext.fault_tolerance;

// NVRxCycleInfo describes a single NVRx cycle for a Slurm job.
// Serialized as JSON and stored in cycle_info.<job_id>.<attempt>.<cycleXX>
// under <base_dir>/nvrx/. The workload reads the current cycle via
// cycle_info.<job_id>.current symlink (see NVRX_CYCLE_INFO_DIR env).
message NVRxCycleInfo {
  // Slurm job id (or Slurm array job id)
  string job_id = 1;

  // Slurm attempt index (SLURM_RESTART_CNT)
  uint32 attempt_index = 2;

  // Cycle number (0-based)
  uint32 cycle_number = 3;

  // Cycle start time, ISO 8601 UTC (e.g. "2025-02-05T14:30:00.123456Z").
  // Python: datetime.fromisoformat(s.replace("Z", "+00:00"))
  string cycle_start_time = 4;

  // Cycle end time, ISO 8601 UTC. Updated when the cycle ends; file is
  // written again (atomically) to persist this.
  string cycle_end_time = 5;

  // Path to the cycle log file for this cycle
  string cycle_log_file = 6;

  // Active nodes used in this cycle, in SLURM node range format (e.g. "node[001-004,007]")
  string active_nodes = 7;

  // Standby (hot spare) nodes in this cycle, in SLURM node range format. Empty if none.
  string standby_nodes = 8;
}
